### PR TITLE
[show][techsupport][multi-ASIC] Add support to collect tech support on multi ASIC platform

### PR DESCRIPTION
**Pull request by smaheshm**
**- What I did**
Add support to collect tech support on multi ASIC platform

Refactored code to put together related commands in a function.

**- How I did it**
Modified generate_dump script to run relevant commands on all namespaces for multi ASIC platform. On single ASIC the commands are run in default (host) namesapce.

**- How to verify it**

Manual verification, and pytest.

```
samaddik@str-serv-acs-14:/var/sonic-mgmt/tests$ pytest show_techsupport/test_techsupport.py --testbed=vms12-t1-lag-1 ....
========= 2 passed in 7049.32 seconds ======


For 2 ASIC output (ASIC hardcoded to 2):

admin@str2-nswitch-acs-4:/var/dump/sonic_dump_str2-nswitch-acs-4_20201105_015555/dump$ ls
APPL_DB.json                             ip.bgp.neighbor.fc00::2.rcv1
APPL_DB.json.1                           ip.bgp.neighbor.fc00::6.adv1
APPL_DB.json.2                           ip.bgp.neighbor.fc00::6.rcv1
ASIC_DB.json                             ip.bgp.neighbor.fc00::a.adv2
ASIC_DB.json.1                           ip.bgp.neighbor.fc00::a.rcv2
ASIC_DB.json.2                           ip.bgp.neighbor.fc00::e.adv2
bgp.ipv6.neighbors.1                     ip.bgp.neighbor.fc00::e.rcv2
bgp.ipv6.neighbors.2                     ip.link
bgp.ipv6.summary.1                       ip.link.1
bgp.ipv6.summary.2                       ip.link.2
bgp.ipv6.table.1                         ip.neigh
bgp.ipv6.table.2                         ip.neigh.1
bgp.neighbors.1                          ip.neigh.2
bgp.neighbors.2                          ip.route
bgp.summary.1                            ip.route.1
bgp.summary.2                            ip.route.2
bgp.table.1                              ip.rule
bgp.table.2                              ip.rule.1
broadcom.knet.debug                      ip.rule.2
broadcom.knet.dma                        ipv6.bgp.neighbor.2603:10e2:400:1::1.adv1
broadcom.knet.dstats                     ipv6.bgp.neighbor.2603:10e2:400:1::1.rcv1
broadcom.knet.link                       ipv6.bgp.neighbor.2603:10e2:400:1::5.adv1
broadcom.knet.rate                       ipv6.bgp.neighbor.2603:10e2:400:1::5.rcv1
broadcom.knet.stats                      ipv6.bgp.neighbor.2603:10e2:400:1::9.adv2
broadcom.nat.egress                      ipv6.bgp.neighbor.2603:10e2:400:1::9.rcv2
broadcom.nat.ingress                     ipv6.bgp.neighbor.2603:10e2:400:1::d.adv2
broadcom.ps                              ipv6.bgp.neighbor.2603:10e2:400:1::d.rcv2
broadcom.soc                             ipv6.bgp.neighbor.fc00::2.adv1
broadcom.version                         ipv6.bgp.neighbor.fc00::2.rcv1
CONFIG_DB.json                           ipv6.bgp.neighbor.fc00::6.adv1
CONFIG_DB.json.1                         ipv6.bgp.neighbor.fc00::6.rcv1
CONFIG_DB.json.2                         ipv6.bgp.neighbor.fc00::a.adv2
COUNTERS_DB.json                         ipv6.bgp.neighbor.fc00::a.rcv2
COUNTERS_DB.json.1                       ipv6.bgp.neighbor.fc00::e.adv2
COUNTERS_DB.json.2                       ipv6.bgp.neighbor.fc00::e.rcv2
df                                       lldpctl

admin@str2-nswitch-acs-4:/var/dump/sonic_dump_str2-nswitch-acs-4_20201105_015555/dump$ head -n 10 ip.neigh.1
10.0.0.13 dev PortChannel0011 lladdr 52:54:00:59:42:aa REACHABLE
240.127.1.1 dev eth0 lladdr 02:42:e9:5b:4f:b0 REACHABLE
10.1.0.6 dev PortChannel4004 lladdr 02:42:f0:7f:01:05 STALE
10.0.0.9 dev PortChannel0008 lladdr 52:54:00:ce:64:93 REACHABLE
10.1.0.4 dev PortChannel4003 lladdr 02:42:f0:7f:01:07 STALE
fe80::9a03:9bff:fe03:2289 dev PortChannel0011 lladdr 98:03:9b:03:22:89 STALE
fe80::9a03:9bff:fe03:2289 dev PortChannel0008 lladdr 98:03:9b:03:22:89 STALE
fc00::a dev PortChannel0008 lladdr 52:54:00:ce:64:93 REACHABLE
2603:10e2:400:1::d dev PortChannel4004 lladdr 02:42:f0:7f:01:05 STALE
fe80::5054:ff:fe59:42aa dev PortChannel0011 lladdr 52:54:00:59:42:aa router STALE

admin@str2-nswitch-acs-4:/var/dump/sonic_dump_str2-nswitch-acs-4_20201105_015555/dump$ head -n 10 ip.neigh.2
10.0.0.39 dev PortChannel0006 lladdr 52:54:00:74:7e:25 REACHABLE
10.0.0.43 dev PortChannel0009 lladdr 52:54:00:c9:2a:65 REACHABLE
10.0.0.45 dev PortChannel0010 lladdr 52:54:00:28:88:af REACHABLE
10.0.0.49 dev PortChannel0013 lladdr 52:54:00:da:b6:77 REACHABLE
10.1.0.10 dev PortChannel4006 lladdr 02:42:f0:7f:01:05 STALE
10.0.0.41 dev PortChannel0007 lladdr 52:54:00:e9:3c:9f REACHABLE
10.0.0.33 dev PortChannel0001 lladdr 52:54:00:ff:ff:06 REACHABLE
10.1.0.8 dev PortChannel4005 lladdr 02:42:f0:7f:01:07 STALE
10.0.0.47 dev PortChannel0012 lladdr 52:54:00:4c:c3:4c REACHABLE
10.0.0.35 dev PortChannel0003 lladdr 52:54:00:6e:90:69 REACHABLE

admin@str2-nswitch-acs-4:/var/dump/sonic_dump_str2-nswitch-acs-4_20201105_015555/dump$ head -n 10 CONFIG_DB.json.2
{
  "ACL_TABLE|DATAACL": {
    "type": "hash",
    "value": {
      "policy_desc": "DATAACL",
      "ports@": "PortChannel0001,PortChannel0003,PortChannel0004,PortChannel0006,PortChannel0007,PortChannel0009,PortChannel0010,PortChannel0012,PortChannel0013,PortChannel0014",
      "stage": "ingress",
      "type": "L3"
    }
  },


```

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

**Comment by smaheshm** [comment]
@abdosi Can you check if any of the commands need to be run on all namespaces. I see 'lldpctl', it doesn't seem to have the option of running on all namespaces. 'sudo ip netns .... ' gives the same output.

File: scripts/generate_dump [hunk 1]
```diff
@@ -113,35 +182,60 @@ save_vtysh() {
 #  None
 ###############################################################################
 save_ip() {
-    local ip_args=$1
+    local ip_args=""$1
```
Comment (smaheshm) [inline scripts/generate_dump:185]: remove

File: scripts/generate_dump [hunk 2]
```diff
@@ -230,6 +396,24 @@ save_platform() {
...
+# Returns:
+#  None
+###############################################################################
+save_platform_info() {
+    save_platform "syseeprom" "platform"
+    save_platform "psustatus" "platform"
+    save_platform "ssdhealth" "platform"
+    save_platform "temperature" "platform"
+    save_platform "fan" "platform"
+
```
Comment (smaheshm) [inline scripts/generate_dump:414]: remove

File: scripts/generate_dump [hunk 3]
```diff
@@ -321,6 +505,148 @@ enable_logrotate() {
...
+        # don't gzip already-gzipped dmesg files :)
+        if [ ! ${file} = "/var/crash/kexec_cmd" -a ! ${file} = "/var/crash/export" ]; then
+            if [[ ${file} == *"kdump."* ]]; then
+                save_file $file kdump false
+            else
+                save_file $file kdump true
+            fi
+        fi
+    done
+
```
Comment (smaheshm) [inline scripts/generate_dump:631]: remove

**Comment by smaheshm** [comment]
🔔

File: scripts/generate_dump [hunk 4]
```diff
@@ -83,8 +84,65 @@ save_cmd() {
...
+    local do_zip=${3:-false}
+    echo ${do_zip}
+
+    # host or default namespace
+    save_cmd "$1" "$2" "$do_zip"
+
+    if [[ ( "$NUM_ASICS" > 1 ) ]] ; then
+      for (( i=1; i<=$NUM_ASICS; i++ ))
+      do
+          ns_cmd="ip netns exec asic$i"
```
Comment (qiluo-msft) [inline scripts/generate_dump:112]: >ns_cmd [](start = 10, length = 6)

Use `sonic-netns-exec` ?

File: scripts/generate_dump [hunk 5]
```diff
@@ -83,8 +84,65 @@ save_cmd() {
...
+# Arguments:
+#  asic_id: (OPTIONAL) ASIC ID
+# Returns:
+#  vtysh namespace option
 ###############################################################################
-# Runs a vtysh command and saves its output to the incrementally built tar.
+get_vtysh_namespace() {
+    local asic_id=${1:-""}
+    local ns=""
+    if [[ ( $asic_id =  "" || $asic_id =  0 ) ]] ; then
```
Comment (qiluo-msft) [inline scripts/generate_dump:133]: >   [](start = 22, length = 2)

Use single blank char instead of two.

File: scripts/generate_dump (hunk 1 above)
Comment (qiluo-msft) [inline scripts/generate_dump:185]: >""$1 [](start = 18, length = 4)

The quotation marks seem weird.

**Comment by rlhui** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Review by qiluo-msft — APPROVED** [review]
LGTM. Suggest also ask @jleveque to review.

**Review by jleveque — APPROVED** [review]

**Review by abdosi — COMMENTED** [review]
Forgot to submit comment so they remain in pending state.

File: scripts/generate_dump [hunk 6]
```diff
@@ -321,6 +505,148 @@ enable_logrotate() {
...
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  ASIC Count
+###############################################################################
+get_asic_count() {
+    local cmd="show platform summary --json | python -c 'import sys, json; \
+               print(json.load(sys.stdin)[\"asic_count\"])'"
```
[the remark this answers is not in the document]
Comment (smaheshm) [reply scripts/generate_dump:641]: The source in both methods is the same so the change should be fine.

File: scripts/generate_dump [hunk 7]
```diff
@@ -321,6 +505,148 @@ enable_logrotate() {
...
+# Globals:
+#  None
+# Arguments:
+#  None
+# Returns:
+#  None
+###############################################################################
+collect_broadcom() {
+    save_cmd "bcmcmd -t5 version" "broadcom.version"
+    save_cmd "bcmcmd -t5 soc" "broadcom.soc"
```
[the remark this answers is not in the document]
Comment (smaheshm) [reply scripts/generate_dump:540]: Sure, I can add this in my next PR.

[the remark this answers is not in the document]
Comment (abdosi) [reply scripts/generate_dump:540]: @smaheshm can we fix this before taking in 201911 and this change for multi-asic

[the remark this answers is not in the document]
Comment (smaheshm) [reply scripts/generate_dump:540]: its coming in #1249

**Pull request merged**